File: echo_server/vonage_video_echo_server.py
# Copyright 2026 Vonage

import argparse
import json
import logging
import os
import queue
import signal
import sys
import threading
from typing import Optional, cast

import vonage_video_connector
from vonage_video_connector.models import (
    AudioData,
    Connection,
    LoggingSettings,
    Publisher,
    PublisherAudioSettings,
    PublisherSettings,
    Session,
    SessionAudioSettings,
    SessionAVSettings,
    SessionSettings,
    SessionVideoPublisherSettings,
    Stream,
    Subscriber,
    SubscriberSettings,
    SubscriberVideoSettings,
    VideoFrame,
    VideoResolution,
)

logger = logging.getLogger(__name__)


def read_session_info(session_info_arg: str) -> dict[str, str]:
    """Read session info from a file path or a direct JSON string.

    Args:
        session_info_arg: Either a path to a JSON file or a raw JSON string.

    Returns:
        Parsed session info dictionary.

    Raises:
        FileNotFoundError: If the path looks like a file but cannot be read.
        json.JSONDecodeError: If the JSON is malformed.
    """
    if os.path.isfile(session_info_arg):
        with open(session_info_arg, "r") as f:
            return cast(dict[str, str], json.load(f))

    return cast(dict[str, str], json.loads(session_info_arg))


class VonageVideoEchoServer:
    def __init__(self, session_info: dict[str, str]) -> None:
        self.client = vonage_video_connector.VonageVideoClient()
        self.audio_thread: Optional[threading.Thread] = None
        self.video_thread: Optional[threading.Thread] = None
        self._stop_event = threading.Event()

        # Simple pass-through queues: frames go in from callbacks, out to publisher.
        # No buffering or gating — echo happens immediately at the natural frame rate.
        self.audio_queue: queue.Queue[AudioData] = queue.Queue()
        self.video_queue: queue.Queue[VideoFrame] = queue.Queue()

        self.is_publishing: bool = False
        self._subscribed_stream_id: Optional[str] = None

        # Build settings
        audio_settings = SessionAudioSettings(sample_rate=48000, number_of_channels=1)
        self.video_resolution = VideoResolution(width=1280, height=720)
        video_publisher_settings = SessionVideoPublisherSettings(
            resolution=self.video_resolution,
            fps=30,
            format="YUV420P",
        )
        session_av_settings = SessionAVSettings(
            audio_publisher=audio_settings, audio_subscribers_mix=audio_settings, video_publisher=video_publisher_settings
        )
        self.session_info = session_info
        self.session_settings = SessionSettings(
            enable_migration=False,
            av=session_av_settings,
            logging=LoggingSettings(level="warn"),
        )
        self.publisher_settings = PublisherSettings(
            name="Video Connector Example Echo Server",
            has_audio=True,
            has_video=True,
            audio_settings=PublisherAudioSettings(enable_stereo_mode=False, enable_opus_dtx=True),
        )

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _copy_audio_frame(audio_data: AudioData) -> AudioData:
        """Create an independent copy of an AudioData frame.

        The SDK may reuse the underlying buffer between callbacks, so we
        must snapshot the sample data before queuing it for later playback.
        """
        buffer_copy = audio_data.sample_buffer.tobytes()
        mem_view = memoryview(buffer_copy).cast("h")
        return AudioData(
            sample_buffer=mem_view,
            sample_rate=audio_data.sample_rate,
            number_of_channels=audio_data.number_of_channels,
            number_of_frames=audio_data.number_of_frames,
        )

    @staticmethod
    def _copy_video_frame(video_frame: VideoFrame) -> VideoFrame:
        """Create an independent copy of a VideoFrame.

        The SDK may reuse the underlying buffer between callbacks, so we
        must snapshot the pixel data before queuing it for later use.
        """
        buffer_copy = bytes(video_frame.frame_buffer)
        mem_view = memoryview(buffer_copy).cast("B")
        return VideoFrame(
            frame_buffer=mem_view,
            resolution=video_frame.resolution,
            format=video_frame.format,
        )

    # ------------------------------------------------------------------
    # Session callbacks
    # ------------------------------------------------------------------

    def on_session_error(self, session: Session, error_description: str, error_code: int) -> None:
        logger.error("Session error: session_id=%s desc=%s code=%d", session.id, error_description, error_code)

    def on_session_connected(self, session: Session) -> None:
        logger.info("Session connected: session_id=%s", session.id)

        own_connection = self.client.get_connection()
        if own_connection:
            logger.info("Own connection: id=%s creation_time=%s", own_connection.id, own_connection.creation_time)

        logger.info("Starting echo server publisher...")
        success = self.client.publish(
            settings=self.publisher_settings,
            on_error_cb=self.on_publisher_error,
            on_stream_created_cb=self.on_stream_created,
            on_stream_destroyed_cb=self.on_stream_destroyed,
        )
        if not success:
            logger.error("Failed to start publishing")

    def on_ready_for_audio(self, session: Session) -> None:
        # Fired when the audio capturer is started (maps to on_ready_to_publish in the
        # C++ SDK, a misleading name).  There is no equivalent callback exposed for
        # video; we start the video echo thread here as well since by this point the
        # publisher pipeline is fully initialised.
        logger.info("Audio system ready, starting echo server: session_id=%s", session.id)
        self.is_publishing = True

        self.audio_thread = threading.Thread(target=self.audio_echo_thread, daemon=False)
        self.audio_thread.start()

        self.video_thread = threading.Thread(target=self.video_echo_thread, daemon=False)
        self.video_thread.start()

    def on_session_disconnected(self, session: Session) -> None:
        logger.info("Session disconnected: session_id=%s", session.id)
        self.stop()

    def on_stream_received(self, session: Session, stream: Stream) -> None:
        logger.info("Stream received: session_id=%s stream_id=%s", session.id, stream.id)
        if self._subscribed_stream_id is not None:
            logger.info("Already subscribed to stream %s, ignoring stream %s", self._subscribed_stream_id, stream.id)
            return
        self._subscribed_stream_id = stream.id
        success = self.client.subscribe(
            stream=stream,
            settings=SubscriberSettings(
                video_settings=SubscriberVideoSettings(
                    preferred_resolution=VideoResolution(width=1280, height=720),
                )
            ),
            on_error_cb=self.on_subscriber_error,
            on_connected_cb=self.on_subscriber_connected,
            on_disconnected_cb=self.on_subscriber_disconnected,
            on_render_frame_cb=self.on_video_frame,
        )
        if not success:
            logger.error("Failed to subscribe to stream %s", stream.id)

    def on_stream_dropped(self, session: Session, stream: Stream) -> None:
        logger.info("Stream dropped: session_id=%s stream_id=%s", session.id, stream.id)
        if self._subscribed_stream_id == stream.id:
            logger.info("Subscribed stream dropped, ready to subscribe to the next available stream")
            self._subscribed_stream_id = None

    def on_connection_created(self, session: Session, connection: Connection) -> None:
        logger.info(
            "Connection created: session_id=%s connection_id=%s creation_time=%s",
            session.id,
            connection.id,
            connection.creation_time,
        )

    def on_connection_dropped(self, session: Session, connection: Connection) -> None:
        own_connection = self.client.get_connection()
        is_own = own_connection and connection.id == own_connection.id
        label = "OWN" if is_own else "OTHER"
        logger.info(
            "Connection dropped [%s]: session_id=%s connection_id=%s",
            label,
            session.id,
            connection.id,
        )

    # ------------------------------------------------------------------
    # Audio processing
    # ------------------------------------------------------------------

    def on_audio_data(self, session: Session, audio_data: AudioData) -> None:
        """Receive audio data and queue it for immediate echo."""
        if not self.is_publishing:
            return
        frame_copy = self._copy_audio_frame(audio_data)
        self.audio_queue.put(frame_copy)

    # ------------------------------------------------------------------
    # Subscriber callbacks
    # ------------------------------------------------------------------

    def on_subscriber_error(self, subscriber: Subscriber, error_description: str, error_code: int) -> None:
        logger.error(
            "Subscriber error: stream_id=%s desc=%s code=%d", subscriber.stream.id, error_description, error_code
        )

    def on_subscriber_connected(self, subscriber: Subscriber) -> None:
        logger.info("Subscriber connected: stream_id=%s", subscriber.stream.id)

    def on_subscriber_disconnected(self, subscriber: Subscriber) -> None:
        logger.info("Subscriber disconnected: stream_id=%s", subscriber.stream.id)

    # ------------------------------------------------------------------
    # Video processing
    # ------------------------------------------------------------------

    def on_video_frame(self, subscriber: Subscriber, video_frame: VideoFrame) -> None:
        """Receive a video frame and queue it for immediate echo.

        Frames at the wrong resolution are discarded — the browser ramps up
        through lower resolutions before settling at 1280x720 and C++ rejects
        mismatched frames anyway.
        """
        if not self.is_publishing:
            return
        if video_frame.resolution.width != self.video_resolution.width or video_frame.resolution.height != self.video_resolution.height:
            return
        frame_copy = self._copy_video_frame(video_frame)
        self.video_queue.put(frame_copy)

    # ------------------------------------------------------------------
    # Publisher callbacks
    # ------------------------------------------------------------------

    def on_publisher_error(self, publisher: Publisher, error_description: str, error_code: int) -> None:
        logger.error(
            "Publisher error: stream_id=%s desc=%s code=%d", publisher.stream.id, error_description, error_code
        )

    def on_stream_created(self, publisher: Publisher) -> None:
        logger.info("Publisher stream created: stream_id=%s", publisher.stream.id)

    def on_stream_destroyed(self, publisher: Publisher) -> None:
        logger.info("Publisher stream destroyed: stream_id=%s", publisher.stream.id)
        self.stop()

    # ------------------------------------------------------------------
    # Echo threads
    # ------------------------------------------------------------------

    def audio_echo_thread(self) -> None:
        """Drain the audio queue and send frames back to the session immediately."""
        logger.info("Audio echo thread started")

        while not self._stop_event.is_set():
            try:
                audio_data = self.audio_queue.get(timeout=0.01)
            except queue.Empty:
                continue
            try:
                self.client.add_audio(audio_data)
            except Exception:
                logger.exception("Error injecting echo audio")

        logger.info("Audio echo thread stopped")

    def video_echo_thread(self) -> None:
        """Drain the video queue and send frames back to the session immediately."""
        logger.info("Video echo thread started")

        while not self._stop_event.is_set():
            try:
                frame = self.video_queue.get(timeout=0.01)
            except queue.Empty:
                continue
            try:
                if not self.client.add_video(frame):
                    logger.warning("add_video() returned False — C++ buffer full, frame dropped")
            except Exception:
                logger.exception("Error injecting video frame")

        logger.info("Video echo thread stopped")

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def stop(self) -> None:
        """Signal the echo threads to stop and wait for them to finish."""
        if self.audio_thread and self.audio_thread.is_alive():
            logger.info("Stopping echo threads...")
            self._stop_event.set()
            self.audio_thread.join(timeout=2.0)
            if self.audio_thread.is_alive():
                logger.warning("Audio echo thread did not stop cleanly")

        if self.video_thread and self.video_thread.is_alive():
            self._stop_event.set()
            self.video_thread.join(timeout=2.0)
            if self.video_thread.is_alive():
                logger.warning("Video echo thread did not stop cleanly")

    def connect(self) -> bool:
        """Connect to the session and start the echo server."""
        logger.info("Connecting to session...")

        success = self.client.connect(
            application_id=self.session_info["apiKey"],
            session_id=self.session_info["sessionId"],
            token=self.session_info["token"],
            session_settings=self.session_settings,
            on_error_cb=self.on_session_error,
            on_connected_cb=self.on_session_connected,
            on_disconnected_cb=self.on_session_disconnected,
            on_connection_created_cb=self.on_connection_created,
            on_connection_dropped_cb=self.on_connection_dropped,
            on_stream_received_cb=self.on_stream_received,
            on_stream_dropped_cb=self.on_stream_dropped,
            on_audio_data_cb=self.on_audio_data,
            on_ready_for_audio_cb=self.on_ready_for_audio,
        )

        if not success:
            logger.error("Failed to connect to session")
            return False

        logger.info("Connected, echo server will echo back any received audio and video")
        return True

    def cleanup(self) -> None:
        """Clean up resources and properly leave the session."""
        self.stop()

        logger.info("Stopping publishing...")
        if not self.client.unpublish():
            logger.warning("Failed to stop publishing")

        if self.client.is_connected():
            logger.info("Disconnecting from session...")
            if not self.client.disconnect():
                logger.warning("Failed to disconnect")

        logger.info("Echo server shut down.")


def main() -> None:
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s [%(levelname)s] %(message)s",
    )

    parser = argparse.ArgumentParser(description="Vonage Video Echo Server")
    parser.add_argument("session_info", help="Path to JSON file with session info or direct JSON string")
    args = parser.parse_args()

    echo_server = VonageVideoEchoServer(read_session_info(args.session_info))

    stop_event = threading.Event()

    def handle_signal(sig: int, frame: object) -> None:
        sig_name = signal.Signals(sig).name
        if stop_event.is_set():
            # Already shutting down — force exit on repeated signal
            logger.warning("Received %s again during cleanup, forcing exit...", sig_name)
            os._exit(1)
        logger.info("Received %s, shutting down echo server...", sig_name)
        stop_event.set()

    def stdin_listener() -> None:
        """Block until Enter is pressed or stdin is closed, then trigger shutdown."""
        try:
            input()
        except EOFError:
            # stdin closed (e.g. non-interactive / detached container) — wait forever
            stop_event.wait()
            return
        logger.info("Enter pressed, shutting down echo server...")
        stop_event.set()

    signal.signal(signal.SIGTERM, handle_signal)
    signal.signal(signal.SIGINT, handle_signal)

    # Daemon thread: listens for Enter key so the server also exits on Enter.
    stdin_thread = threading.Thread(target=stdin_listener, daemon=True)
    stdin_thread.start()

    try:
        if not echo_server.connect():
            sys.exit(1)

        logger.info("Echo server running. Press Enter or Ctrl+C to stop...")
        stop_event.wait()

    finally:
        echo_server.cleanup()


if __name__ == "__main__":
    main()

File: nodejs/echo_server/vonage_video_echo_server.ts
#!/usr/bin/env node


import * as fs from 'fs';

import {
  VonageVideoClient,
  type Session,
  type Stream,
  type Connection,
  type Subscriber,
  type Publisher,
  type AudioData,
  type SessionSettings,
  type PublisherSettings,
  type SampleRate,
  type NumberOfChannels,
  type VideoFrame,
} from '@vonage/video-connector';

function readSessionInfo(sessionInfoArg: string): { apiKey: string; sessionId: string; token: string } {
  if (fs.existsSync(sessionInfoArg) && fs.statSync(sessionInfoArg).isFile()) {
    return JSON.parse(fs.readFileSync(sessionInfoArg, 'utf8'));
  }
  return JSON.parse(sessionInfoArg);
}

// Scale a YUV420P video frame to the given width and height, preserving aspect ratio,
// centering the result and converting it to grayscale.
// Returns a new VideoFrame object with the scaled data.
function scaleGrayVideoFrame(frame: VideoFrame, width: number, height: number): VideoFrame {
  if (frame.format === 'YUV420P') {
    const yuvData = frame.data;
    const srcWidth = frame.width;
    const srcHeight = frame.height;

    // Compute the largest scaled size that fits within the target while
    // preserving the source aspect ratio (fit-inside / letterbox strategy).
    const srcAspect = srcWidth / srcHeight;
    const dstAspect = width / height;

    let scaledWidth: number;
    let scaledHeight: number;
    if (srcAspect > dstAspect) {
      // Source is wider than destination — fit to width, letterbox top/bottom
      scaledWidth = width;
      scaledHeight = Math.round(width / srcAspect);
    } else {
      // Source is taller than destination — fit to height, pillarbox left/right
      scaledHeight = height;
      scaledWidth = Math.round(height * srcAspect);
    }

    // YUV420P requires even dimensions for correct chroma plane addressing
    scaledWidth  = scaledWidth  & ~1;
    scaledHeight = scaledHeight & ~1;

    // Even-aligned centering offsets so chroma samples stay on 2×2 boundaries
    const xOffset = (Math.floor((width  - scaledWidth)  / 2)) & ~1;
    const yOffset = (Math.floor((height - scaledHeight) / 2)) & ~1;

    // YUV420P layout:
    // Y plane  : width * height bytes
    // U plane  : (width/2) * (height/2) bytes
    // V plane  : (width/2) * (height/2) bytes
    const srcYPlaneSize = srcWidth * srcHeight;
    const dstYPlaneSize = width * height;
    const dstUPlaneSize = (width / 2) * (height / 2);

    // Initialise the output buffer with YUV black:
    const yBlack = 16; // Y = 16 (limited-range black)
    const uvNeutral = 128; // U = V = 128 (grayscale)
    const scaledData = Buffer.alloc(width * height * 3 / 2);
    scaledData.fill(yBlack, 0, dstYPlaneSize);
    scaledData.fill(uvNeutral, dstYPlaneSize, dstYPlaneSize + dstUPlaneSize * 2);

    // Inverse scale factors for nearest-neighbor source lookup
    const xScale = srcWidth / scaledWidth;
    const yScale = srcHeight/ scaledHeight;

    // Scale Y plane into the centred region using nearest-neighbor.
    // U and V are set to 128 (already done above) — grayscale has no chroma.
    for (let y = 0; y < scaledHeight; y++) {
      for (let x = 0; x < scaledWidth; x++) {
        const srcX = Math.floor(x * xScale);
        const srcY = Math.floor(y * yScale);
        scaledData[(y + yOffset) * width + (x + xOffset)] = yuvData[srcY * srcWidth + srcX];
      }
    }

    return {
      data: scaledData,
      width: width,
      height: height,
      format: 'YUV420P',
    };
  } else {
    throw new Error(`Unsupported video format for scaling: ${frame.format}`);
  }
}

class VonageVideoEchoServer {
  static readonly DEFAULT_SAMPLE_RATE: SampleRate = 48000;
  static readonly DEFAULT_NUMBER_OF_CHANNELS: NumberOfChannels = 2;
  static readonly AUDIO_TICK_INTERVAL = 10; // ms
  static readonly VIDEO_PUBLISH_WIDTH =  640; // pixels
  static readonly VIDEO_PUBLISH_HEIGHT =  480; // pixels
  static readonly VIDEO_PUBLISH_FPS =  30; // ms
  static readonly VIDEO_TICK_INTERVAL = 1000 / VonageVideoEchoServer.VIDEO_PUBLISH_FPS; // ms

  private sessionInfo: { apiKey: string; sessionId: string; token: string };
  private client: VonageVideoClient;
  private audioTimer: NodeJS.Timeout | null;
  private audioQueue: AudioData[];
  private videoTimer: NodeJS.Timeout | null;
  private videoQueue: VideoFrame[];
  private isPublishing: boolean;
  private streams: Map<string, Stream>;
  private echoedVideoStreamId: string | null;
  private sessionSettings: SessionSettings;
  private publisherSettings: PublisherSettings;

  constructor(sessionInfo: { apiKey: string; sessionId: string; token: string }) {
    this.sessionInfo = sessionInfo;
    this.client = new VonageVideoClient();

    this.audioTimer = null;
    this.audioQueue = [];
    this.videoTimer = null;
    this.videoQueue = [];
    this.isPublishing = false;
    this.streams = new Map<string, Stream>();
    this.echoedVideoStreamId = null;

    this.sessionSettings = {
      enableMigration: false,
      av: {
        audioPublisher: { sampleRate: VonageVideoEchoServer.DEFAULT_SAMPLE_RATE, channels: VonageVideoEchoServer.DEFAULT_NUMBER_OF_CHANNELS },
        audioSubscribersMix: { sampleRate: VonageVideoEchoServer.DEFAULT_SAMPLE_RATE, channels: VonageVideoEchoServer.DEFAULT_NUMBER_OF_CHANNELS },
        videoPublisher: { 
          width: VonageVideoEchoServer.VIDEO_PUBLISH_WIDTH,
          height: VonageVideoEchoServer.VIDEO_PUBLISH_HEIGHT,
          fps: VonageVideoEchoServer.VIDEO_PUBLISH_FPS,
          format: 'YUV420P'
        },
      },
      logging: { level: 'DEBUG' },
      // logging: { level: 'WARN' },
    };

    this.publisherSettings = {
      name: 'Video Connector Example Echo Server',
      hasAudio: true,
      hasVideo: true,
      audioSettings: {
        enableStereoMode: false,
        enableOpusDtx: true,
      },
    };
  }

  log(level: string, msg: string, ...args: unknown[]): void {
    const ts = new Date().toISOString();
    const line = `${ts} [${level.toUpperCase()}] ${msg}`;
    if (level === 'error') {
      console.error(line, ...args);
    } else if (level === 'warn') {
      console.warn(line, ...args);
    } else {
      console.log(line, ...args);
    }
  }

  onSessionError(session: Session, errorDescription: string, errorCode: number): void {
    this.log('error', `Session error: session_id=${session?.sessionId} desc=${errorDescription} code=${errorCode}`);
  }

  onReadyForAudio(session: Session): void {
    this.log('info', `Audio system ready, echo server is now active: session_id=${session?.sessionId}`);
    this.isPublishing = true;

    if (!this.audioTimer) {
      this.audioTimer = setInterval(() => this.audioEchoTick(), VonageVideoEchoServer.AUDIO_TICK_INTERVAL);
    }
  }

  onSessionDisconnected(session: Session): void {
    this.log('info', `Session disconnected: session_id=${session?.sessionId}`);
    this.stop();
  }

  async onStreamReceived(session: Session, stream: Stream): Promise<void> {
    this.log('info', `Stream received: session_id=${session?.sessionId} stream_id=${stream?.streamId}`);
    if (!this.videoTimer) {
      this.videoTimer = setInterval(() => this.videoEchoTick(), VonageVideoEchoServer.VIDEO_TICK_INTERVAL);
    }

    this.streams.set(stream?.streamId, stream);
    if(!this.echoedVideoStreamId) {
      this.echoedVideoStreamId = stream?.streamId;
    }

    const subscriber = await this.client.subscribe(
      stream,
      {
        onError: this.onSubscriberError.bind(this),
        onDisconnected: this.onSubscriberDisconnected.bind(this),
        onRenderFrame: (subscriber: Subscriber, frame: VideoFrame) => {
          if( subscriber.stream?.streamId === this.echoedVideoStreamId) {
            this.videoQueue.push(frame);
          }
        },
      }
    );

    if (!subscriber) {
      this.log('error', `Failed to subscribe to stream ${stream?.streamId}`);
    }

    this.log('info', `Subscribed to stream: session_id=${session?.sessionId} stream_id=${stream?.streamId}`);
  }

  onStreamDropped(session: Session, stream: Stream): void {
    this.log('info', `Stream dropped: session_id=${session?.sessionId} stream_id=${stream?.streamId}`);

    this.streams.delete(stream?.streamId);
    if(this.echoedVideoStreamId && stream?.streamId === this.echoedVideoStreamId) {
      this.echoedVideoStreamId = this.streams.size > 0 ? Array.from(this.streams.keys())[0] : null;
    }
    if (this.streams.size === 0) {
      this.log('info', 'No more streams, stopping video echo thread...');
      clearInterval(this.videoTimer);
      this.videoTimer = null;
    }
  }

  onConnectionCreated(session: Session, connection: Connection): void {
    this.log(
      'info',
      `Connection created: session_id=${session?.sessionId} connection_id=${connection?.connectionId} creation_time=${connection?.creationTime}`
    );
  }

  onConnectionDropped(session: Session, connection: Connection): void {
    const ownConnection = this.client.getConnection();
    const isOwn = ownConnection && connection?.connectionId === ownConnection.connectionId;
    const label = isOwn ? 'OWN' : 'OTHER';
    this.log('info', `Connection dropped [${label}]: session_id=${session?.sessionId} connection_id=${connection?.connectionId}`);
  }

  onSessionAudioData(_session: Session, audioData: AudioData): void {
    if (!this.isPublishing) return;
    this.audioQueue.push(audioData);
  }

  onSubscriberError(subscriber: Subscriber, errorDescription: string, errorCode: number): void {
    this.log('error', `Subscriber error: stream_id=${subscriber?.stream?.streamId} desc=${errorDescription} code=${errorCode}`);
  }

  onSubscriberDisconnected(subscriber: Subscriber): void {
    this.log('info', `Subscriber disconnected: stream_id=${subscriber?.stream?.streamId}`);
  }

  onPublisherError(publisher: Publisher, errorDescription: string, errorCode: number): void {
    this.log('error', `Publisher error: stream_id=${publisher?.stream?.streamId} desc=${errorDescription} code=${errorCode}`);
  }

  onStreamDestroyed(publisher: Publisher): void {
    this.log('info', `Publisher stream destroyed: stream_id=${publisher?.stream?.streamId}`);
    this.stop();
  }

  audioEchoTick() {
    if (this.audioQueue.length === 0) return;

    const audioData = this.audioQueue.shift();
    if (audioData === undefined) return;

    try {
      this.client.addAudio(audioData);
    } catch (error) {
      this.log('error', 'Error injecting echo audio', error);
    }
  }

  videoEchoTick() {
    if (this.videoQueue.length === 0) return;

    const videoFrame = this.videoQueue.shift();
    if (videoFrame === undefined) return;

    try {
      // Scale the video frame to the publisher's resolution before sending it back and take out the color information to make it grayscale
      const scaledFrame: VideoFrame = scaleGrayVideoFrame(videoFrame, VonageVideoEchoServer.VIDEO_PUBLISH_WIDTH, VonageVideoEchoServer.VIDEO_PUBLISH_HEIGHT);
      this.client.addVideo(scaledFrame);
    } catch (error) {
      this.log('error', 'Error injecting echo video', error);
    }
  }

  stop() {
    if (this.audioTimer) {
      this.log('info', 'Stopping audio echo thread...');
      clearInterval(this.audioTimer);
      this.audioTimer = null;
    }
    if (this.videoTimer) {
      this.log('info', 'Stopping video echo thread...');
      clearInterval(this.videoTimer);
      this.videoTimer = null;
    }
  }

  async connect() {
    this.log('info', 'Connecting to session...');

    const session = await this.client.connect(
      this.sessionInfo.apiKey,
      this.sessionInfo.sessionId,
      this.sessionInfo.token,
      {
        sessionSettings: this.sessionSettings,
        onError: this.onSessionError.bind(this),
        onDisconnected: this.onSessionDisconnected.bind(this),
        onConnectionCreated: this.onConnectionCreated.bind(this),
        onConnectionDropped: this.onConnectionDropped.bind(this),
        onStreamReceived: this.onStreamReceived.bind(this),
        onStreamDropped: this.onStreamDropped.bind(this),
        onAudioData: this.onSessionAudioData.bind(this),
        onReadyForAudio: this.onReadyForAudio.bind(this),
    });

    if (!session) {
      this.log('error', 'Failed to connect to session');
      return false;
    }

    const own_connection = this.client.getConnection();
    if (own_connection) {
        this.log('info', `Own connection: id=${own_connection.connectionId} creation_time=${own_connection.creationTime}`);
    }

    this.log('info', `Connected to session: session_id=${session?.sessionId}, echo server will echo back any received audio.`);
    return true;
  }

  async publish() {
    this.log('info', 'Starting publishing...');
    const publisher = await this.client.publish(
      {
        settings: this.publisherSettings,
        onError: this.onPublisherError.bind(this),
        onStreamDestroyed: this.onStreamDestroyed.bind(this),
      }
    );

    if (!publisher) {
      this.log('error', 'Failed to start publishing');
      return false;
    }

    return true;
  } 
  

  cleanup() {
    this.stop();

    this.log('info', 'Stopping publishing...');
    this.client.unpublish().catch(() => {
      this.log('warn', 'Failed to stop publishing');
    });

    const connected = this.client.isConnected();
    if (connected) {
      this.log('info', 'Disconnecting...');
      this.client.disconnect().catch(() => {
        this.log('warn', 'Failed to disconnect');
      });
    }
  }
}

function waitForShutdownSignal(): Promise<NodeJS.Signals> {
  return new Promise<NodeJS.Signals>((resolve) => {
    // Keep the event loop alive while waiting; signal listeners alone do not.
    const keepAlive = setInterval(() => {}, 1 << 30);

    const onSignal = (signal: NodeJS.Signals) => {
      clearInterval(keepAlive);
      resolve(signal);
    };

    process.once('SIGINT', onSignal);
    process.once('SIGTERM', onSignal);
  });
}

async function main() {

  const sessionInfoArg: string = process.argv[2];
  if (!sessionInfoArg) {
    console.error('Usage: node vonage_video_echo_server.js <session.json|json-string>');
    process.exit(1);
  }

  let sessionInfo: { apiKey: string; sessionId: string; token: string };
  try {
    sessionInfo = readSessionInfo(sessionInfoArg);
  } catch (error) {
    console.error('Invalid session info. Provide a JSON file path or a JSON string.', error instanceof Error ? error.message : error);
    process.exit(1);
  }

  const echoServer = new VonageVideoEchoServer(sessionInfo);

  try {
    if (!await echoServer.connect()) {
      process.exit(1);
    }

    if (!await echoServer.publish()) {
      process.exit(1);
    }

    echoServer.log('info', 'Echo server running. Press Ctrl+C to stop...');
    const signal = await waitForShutdownSignal();
    echoServer.log('info', `Received ${signal}, shutting down echo server...`);
  } catch (error) {
    echoServer.log('error', 'Unhandled runtime error', error);
    process.exitCode = 1;
  } finally {
    echoServer.cleanup();
  }
}

main();
